File: frog/src/main/java/yuluyao/frog/drag/DragCallback.java
package yuluyao.frog.drag;

import android.os.Build;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import yuluyao.frog.FrogAdapter;

import java.util.Collections;
import java.util.List;

/**
 * 描 述：
 * 作 者：Vegeta Yu
 * 时 间：2017/9/14 17:37
 */
public class DragCallback extends ItemTouchHelper.Callback {

  @Override
  public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
    int dragFlags;
    int swipeFlags;
    dragFlags =
        ItemTouchHelper.UP | ItemTouchHelper.DOWN | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
    swipeFlags = 0;
    return makeMovementFlags(dragFlags, swipeFlags);
  }

  @Override public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder,
      RecyclerView.ViewHolder target) {
    int fromPosition = viewHolder.getAdapterPosition();//得到拖动ViewHolder的position
    int toPosition = target.getAdapterPosition();//得到目标ViewHolder的position
    List list = ((FrogAdapter) recyclerView.getAdapter()).getData();
    if (fromPosition < toPosition) {
      for (int i = fromPosition; i < toPosition; i++) {
        if (list != null) {
          Collections.swap(list, i, i + 1);
        }
      }
    } else {
      for (int i = fromPosition; i > toPosition; i--) {
        if (list != null) {
          Collections.swap(list, i, i - 1);
        }
      }
    }
    recyclerView.getAdapter().notifyItemMoved(fromPosition, toPosition);
    return true;
  }

  @Override public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {

  }

  @Override public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
    super.onSelectedChanged(viewHolder, actionState);
    switch (actionState) {
      case ItemTouchHelper.ACTION_STATE_IDLE:
        break;
      case ItemTouchHelper.ACTION_STATE_SWIPE:
        break;
      case ItemTouchHelper.ACTION_STATE_DRAG:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
          viewHolder.itemView.setElevation(6f);
        }
        break;
    }
  }

  @Override public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
    super.clearView(recyclerView, viewHolder);
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      viewHolder.itemView.setElevation(0f);
    }
  }
}

File: frogx/src/main/java/yuluyao/frog/drag/DragCallback.kt
package yuluyao.frog.drag

import android.content.res.Resources
import android.os.Build

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import yuluyao.frog.FrogAdapter

import java.util.Collections

/**
 * 描 述：
 * 作 者：Vegeta Yu
 * 时 间：2017/9/14 17:37
 */
class DragCallback : ItemTouchHelper.Callback() {

  override fun getMovementFlags(recyclerView: RecyclerView,
                                viewHolder: RecyclerView.ViewHolder): Int {
    val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN or ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
    val swipeFlags = 0
    return makeMovementFlags(dragFlags, swipeFlags)
  }

  override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder,
                      target: RecyclerView.ViewHolder): Boolean {
    //得到拖动ViewHolder的position
    val fromPosition = viewHolder.adapterPosition
    //得到目标ViewHolder的position
    val toPosition = target.adapterPosition
    val list = (recyclerView.adapter as FrogAdapter<*>).data
    if (fromPosition < toPosition) {
      for (i in fromPosition until toPosition) {
        Collections.swap(list, i, i + 1)
      }
    } else {
      for (i in fromPosition downTo toPosition + 1) {
        Collections.swap(list, i, i - 1)
      }
    }
    recyclerView.adapter!!.notifyItemMoved(fromPosition, toPosition)
    return true
  }

  override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {

  }

  override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
    super.onSelectedChanged(viewHolder, actionState)
    when (actionState) {
      ItemTouchHelper.ACTION_STATE_IDLE -> {
      }
      ItemTouchHelper.ACTION_STATE_SWIPE -> {
      }
      ItemTouchHelper.ACTION_STATE_DRAG -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
        viewHolder!!.itemView.elevation = 4 * Resources.getSystem().displayMetrics.density
      }
    }
  }

  override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
    super.clearView(recyclerView, viewHolder)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      viewHolder.itemView.elevation = 0f
    }
  }
}
